### PR TITLE
HyperKit instance plugin

### PR DESCRIPTION
**Pull request by chungers**
This is an implementation of an instance plugin provisioning vm guest using HyperKit:

### Instance Plugin
  + It listens on a tcp port so that controllers running in Docker container can connect to it as a network service
  + It provisions vm guest instance using images built from LinuxKit
  + The instance Properties is simply the `HyperKit` struct in the hyperkit golang package.  This allows full fidelity in configuration of the hyperkit vm guest.
  + The plugin supports retrieving full properties of the instance by reading the `hyperkit.json` file in the instance state directory.
  + The plugin no longer removes the guest instance when it's not running.  

### Playbook
   + Added new `mac` option in the `docs/playbooks/intro` which can be added via the command line `infrakit playbook add <name> file://$PWD/docs/playbooks/intro/index.ikb`
   + A new command for starting the plugin is added (see `infrakit playbook mac -h`
   + A new command for provisioning instance using an YAML template of the input is included  (see `docs/playbooks/intro/macosx/provision-instance.ikt`
   + A new option is added so that the user can interactively start up the hyperkit plugin via `infrakit playbook <the_name_used_when_adding_playbook> start-infrakit`

Here is an example for creating a new command in the infrakit CLI that will get the required data from flags or prompting the user. 

```
{{/* Input to create instance using the HyperKit instance plugin */}}
{{/* =% instanceProvision "instance-hyperkit" true  %= */}}

{{ $user := flag "user" "string" "username" | prompt "Please enter your user name:" "string" (env "USER")}}
{{ $mobyPath := flag "moby-dirh" "string" "Moby directory" | prompt "Moby directory?" "string" (cat (env "HOME") "/projects/src/github.com/linuxkit/linuxkit" | nospace) }}
{{ $bundle := flag "bundle" "string" "Bundle prefix" | prompt "Bundle prefix?" "string"}}
{{ $privateIP := flag "private-ip" "string" "Private IP" | prompt "Private IP address (IPv4)?" "string" "" nil }}
LogicalID: {{ $privateIP }}
Tags:
  infrakit.created: {{ now | htmlDate }}
  infrakit.user: {{ $user }}

Init: |
  #!/bin/bash
  echo Hello world!

Properties:
  kernel: {{ cat $mobyPath "/" $bundle "-bzImage" | nospace }}
  initrd: {{ cat $mobyPath "/" $bundle "-initrd.img" | nospace }}
  cpus: 1
  memory: 256
  disk_size: 100
  cmdline: console=ttyS0 page_poison=1

``` 

### Using
0. Add a playbook
```shell
$ infrakit playbook add INTRO file://`pwd`/docs/playbooks/intro/index.ikb
```

1. Starting the daemon:
```shell
$ infrakit playbook INTRO mac start-daemon
Start HYPERKIT plugin? [no]: y
Starting HYPERKIT plugin.  This must be running on the Mac as a daemon and not as a container
time="2017-04-23T23:45:57-07:00" level=info msg="Listening at: localhost:24865, discoverable at /Users/davidchung/.infrakit/plugins/instance-hyperkit.listen" 
time="2017-04-23T23:45:57-07:00" level=info msg="PID file at /Users/davidchung/.infrakit/plugins/instance-hyperkit.pid" 
```

2. In another terminal, creating an instance
```shell
$ infrakit playbook INTRO mac provision-instance
Please enter your user name: [davidchung]: 
Moby directory? [/Users/davidchung/projects/src/github.com/linuxkit/linuxkit]: 
Bundle prefix? : sshd
Private IP address (IPv4)? : 
infrakit-143783457
```
3. See the instance
```shell
$ infrakit instance --name instance-hyperkit describe -pry
- ID: infrakit-143783457
  LogicalID: infrakit-143783457
  Properties:
    arguments:
    - -A
    - -u
    - -F
    - /Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/hyperkit.pid
    - -c
    - "1"
    - -m
    - 256M
    - -s
    - 0:0,hostbridge
    - -s
    - 1:0,virtio-vpnkit,path=auto
    - -s
    - 2:0,virtio-blk,/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/disk.img
    - -s
    - 4,ahci-cd,/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/data.iso
    - -s
    - 10,virtio-rnd
    - -s
    - 31,lpc
    - -l
    - com1,autopty=/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/tty,log=/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/console-ring
    - -f
    - kexec,/Users/davidchung/projects/src/github.com/linuxkit/linuxkit/sshd-bzImage,/Users/davidchung/projects/src/github.com/linuxkit/linuxkit/sshd-initrd.img,earlyprintk=serial
      console=ttyS0 page_poison=1
    cmdline: /usr/local/bin/hyperkit -A -u -F /Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/hyperkit.pid
      -c 1 -m 256M -s 0:0,hostbridge -s 1:0,virtio-vpnkit,path=auto -s 2:0,virtio-blk,/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/disk.img
      -s 4,ahci-cd,/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/data.iso
      -s 10,virtio-rnd -s 31,lpc -l com1,autopty=/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/tty,log=/Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/console-ring
      -f kexec,/Users/davidchung/projects/src/github.com/linuxkit/linuxkit/sshd-bzImage,/Users/davidchung/projects/src/github.com/linuxkit/linuxkit/sshd-initrd.img,earlyprintk=serial
      console=ttyS0 page_poison=1
    console: 1
    cpus: 1
    disk: /Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/disk.img
    disk_size: 100
    hyperkit: /usr/local/bin/hyperkit
    initrd: /Users/davidchung/projects/src/github.com/linuxkit/linuxkit/sshd-initrd.img
    iso: /Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457/data.iso
    kernel: /Users/davidchung/projects/src/github.com/linuxkit/linuxkit/sshd-bzImage
    memory: 256
    pid: 96767
    running: true
    state_dir: /Users/davidchung/.infrakit/hyperkit-vms/infrakit-143783457
    uuid: ""
    vpnkit_key: ""
    vpnkit_sock: auto
    vsock: false
  Tags:
    infrakit.created: 2017-04-23
    infrakit.user: davidchung
```

@rneugeba

### CODE REVIEW

File: cmd/instance/hyperkit/main.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+		Short: "HyperKit instance plugin",
+	}
+
+	defaultVMDir := filepath.Join(getHome(), ".infrakit/hyperkit-vms")
+
+	name := cmd.Flags().String("name", "instance-hyperkit", "Plugin name to advertise for discovery")
+	logLevel := cmd.Flags().Int("log", cli.DefaultLogLevel, "Logging level. 0 is least verbose. Max is 5")
+
+	vmDir := cmd.Flags().String("vm-dir", defaultVMDir, "Directory where to store VM state")
+	hyperkitCmd := cmd.Flags().String("hyperkit", "/usr/local/bin/hyperkit", "Path to HyperKit executable")
```
Comment (rneugeba) [inline cmd/instance/hyperkit/main.go]: I think it's best to keep this empty and let the hyperkit go binding try common locations

File: pkg/plugin/instance/hyperkit/instance.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path"
+
+	logutil "github.com/docker/infrakit/pkg/log"
+	"github.com/docker/infrakit/pkg/spi/instance"
+	"github.com/docker/infrakit/pkg/types"
+	hyperkit_go "github.com/moby/hyperkit/go"
```
Comment (rneugeba) [inline pkg/plugin/instance/hyperkit/instance.go]: this imports as `hyperkit`. Any reason to rename it to `hyperkit_go`?

File: pkg/plugin/instance/hyperkit/instance.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+		// non-zero, we place the disk in a special directory
+		// so it persists across reboots.
+		if guest.DiskSize > 0 {
+			guest.DiskImage = path.Join(p.DiskDir, logicalID+".img")
+		}
+	}
+
+	// if there's init then build an iso image of that
+	if spec.Init != "" {
+
```
Comment (rneugeba) [inline pkg/plugin/instance/hyperkit/instance.go]: this line can probably be removed

File: pkg/plugin/instance/hyperkit/instance.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+	_, err := os.Stat(instanceDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return errors.New("Instance does not exist")
+		}
+	}
+
+	h, err := hyperkit_go.FromState(instanceDir)
+	if err != nil {
+		return err
```
Comment (rneugeba) [inline pkg/plugin/instance/hyperkit/instance.go]: should this be a `log.Crit`?

File: pkg/plugin/instance/hyperkit/instance.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+			logicalID = &lid
+
+			desc := instance.Description{
+				ID:        id,
+				LogicalID: logicalID,
+				Tags:      instanceTags,
+			}
+
+			if properties {
+
```
Comment (rneugeba) [inline pkg/plugin/instance/hyperkit/instance.go]: this line can be removed

File: docs/playbooks/intro/index.ikb [hunk 6]
```diff
@@ -12,3 +12,4 @@ events         : events/index.ikb
 
 aws : aws/index.ikb
 gcp : gcp/index.ikb
+mac : macosx/index.ikb
```
Comment (rneugeba) [inline docs/playbooks/intro/index.ikb]: `macosx` is a little odd. I'd either call it `darwin` (as in GOOS) or `macos` (as in the new official name macOS)

File: docs/playbooks/intro/macosx/provision-instance.ikt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+{{/* Input to create instance using the HyperKit instance plugin */}}
+{{/* =% instanceProvision "instance-hyperkit" true  %= */}}
+
+{{ $user := flag "user" "string" "username" | prompt "Please enter your user name:" "string" (env "USER")}}
+{{ $mobyPath := flag "moby-dirh" "string" "Moby directory" | prompt "Moby directory?" "string" (cat (env "HOME") "/projects/src/github.com/linuxkit/linuxkit" | nospace) }}
```
Comment (rneugeba) [inline docs/playbooks/intro/macosx/provision-instance.ikt]: what does `moby` refer to here? The name is confusing. Should this be `linuxkitPath` or `imagePath`?

**Comment by rneugeba** [comment]
Overall, this looks good. Thanks for porting it over from LinuxKit. I added a few minor niggles/comments

**Comment by chungers** [comment]
@rneugeba  - thanks for the review.  This instance plugin has been verified with the linuxkit sshd example.

**Pull request merged**